src/components/ComputeModal.js

```diff
@@ -309,24 +277,59 @@ export const ComputeModalBase = ({
     if (shouldDeletePersistentDisk && !hasAttachedDisk()) {
       await Ajax().Disks.disk(googleProject, currentPersistentDiskDetails.name).delete()
     }
-    if (shouldUpdateRuntime) {
-      await Ajax().Runtimes.runtime(googleProject, currentRuntimeDetails.runtimeName).update({
-        runtimeConfig,
-        autopauseThreshold: computeConfig.autopauseThreshold
-      })
-    }
-    if (shouldCreateRuntime) {
-      await Ajax().Runtimes.runtime(googleProject, Utils.generateRuntimeName()).create({
-        runtimeConfig,
-        autopauseThreshold: computeConfig.autopauseThreshold,
-        toolDockerImage: desiredRuntime.toolDockerImage,
-        labels: {
-          saturnWorkspaceNamespace: namespace,
-          saturnWorkspaceName: name
-        },
-        customEnvironmentVariables: customEnvVars,
-        ...(desiredRuntime.jupyterUserScriptUri ? { jupyterUserScriptUri: desiredRuntime.jupyterUserScriptUri } : {})
-      })
+
+    if (shouldUpdateRuntime || shouldCreateRuntime) {
+      const runtimeConfig = desiredRuntime && {
+        cloudService: desiredRuntime.cloudService,
+        ...(desiredRuntime.cloudService === cloudServices.GCE ? {
+          zone: desiredRuntime.zone.toLowerCase(),
+          machineType: desiredRuntime.machineType || defaultGceMachineType,
+          ...(computeConfig.gpuEnabled && { gpuConfig: { gpuType: computeConfig.gpuType, numOfGpus: computeConfig.numGpus } })
+        } : {
+          region: desiredRuntime.region.toLowerCase(),
+          masterMachineType: desiredRuntime.masterMachineType || defaultDataprocMachineType,
+          masterDiskSize: desiredRuntime.masterDiskSize,
+          numberOfWorkers: desiredRuntime.numberOfWorkers,
+          componentGatewayEnabled: desiredRuntime.componentGatewayEnabled,
+          ...(desiredRuntime.numberOfWorkers && {
+            numberOfPreemptibleWorkers: desiredRuntime.numberOfPreemptibleWorkers,
+            workerMachineType: desiredRuntime.workerMachineType,
+            workerDiskSize: desiredRuntime.workerDiskSize
+          })
+        })
+      }
+
+      if (shouldUpdateRuntime) {
+        const updateRuntimeConfig = _.merge(shouldUpdatePersistentDisk ? { diskSize: desiredPersistentDisk.size } : {}, runtimeConfig)
```

Comment:
Does this need to be a deep merge or would shallow merge suffice (e.g. using the spread operator)?

Reply:
The spread operator would work, but it looks a bit cleaner this way imo. I'm not tied to it though

Reply:
I like the way it looks too; just thought we could eliminate unnecessary operations and adhere to the rule of least power but I don't feel strongly.